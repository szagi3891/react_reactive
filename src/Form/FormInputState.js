//@flow
import { Value, ValueComputed, combineValue } from 'computed-values';

export default class FormInputState {
    _value = new Value('');
    _isVisit = new Value(false);
                                                    //dla widoku
    value$ = this._value.asComputed();
    error$: ValueComputed<string | null>;
                                                    //dla wy≈ºszego stanu
    data$: ValueComputed<string | null>;

    constructor(errorMessage: string, fnValidator: (value: string) => bool) {

        const errorInput$ = this.value$.map((input: string): bool => !fnValidator(input));

        this.error$ = combineValue(
            errorInput$,
            this._isVisit.asComputed(),
            (error: bool, isVisit: bool): string | null =>
                (error && isVisit) ? errorMessage : null
        );

        this.data$ = combineValue(
            errorInput$,
            this.value$,
            (error: bool, value: string): string | null => {
                if (error) {
                    return null;
                }

                return value;
            }
        )
    }

    onChange = (event: SyntheticInputEvent<>) => {
        this._value.setValue(event.target.value);
    }
 
    onBlur = () => {
        this._isVisit.setValue(true);
    }
}
