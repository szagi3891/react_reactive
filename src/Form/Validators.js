//@flow

export const isNumber = (text: string): bool => parseInt(text, 10).toString() === text;

export const isGrunwald = (text: string): bool => text === "1410";

const isHexDigit = (digit: string): bool => {
    if (isNumber(digit)) {
        return true;
    }

    const letter = digit.toLowerCase();
    return ["a", "b", "c", "d", "e", "f"].includes(letter);
}

export const isHex = (text: string): bool => {
    if (text.length === 0) {
        return false;
    }

    const maxIndex = text.length - 1;
    
    for (let i = 0; i <= maxIndex; i++) {
        if (isHexDigit(text[i]) === false) {
            return false;
        }
    }

    return true;
}

export default {
    isNumber,
    isGrunwald,
    isHex
};
