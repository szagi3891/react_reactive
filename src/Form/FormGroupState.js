//@flow
import { ValueObservable, Subject, Observable } from '../Lib/Reactive';
import FormInputState from './FormInputState';

const filterNull = (list: Array<string | null>): Array<string> => {
    const out = [];
    for (const item of list) {
        if (item !== null) {
            out.push(item);
        }
    }
    return out;
};

type InputConfig = {|
    key: string,
    label: string,
    state: FormInputState
|};

export default class FormGroupState {

    inputs: Array<InputConfig>;
    errors$: ValueObservable<Array<string>>;
    data$: ValueObservable<Array<string>>;
                                                    //strumie≈Ñ z poprawnie zwalidowanymi danymi formularza
    submitData$: Observable<Array<string>>;

    send: Subject<void> = new Subject();
    send$ = this.send.asObservable();

    onSend = () => {
        this.send.next();
    }

    constructor(inputsConfig: Array<InputConfig>) {
        this.inputs = inputsConfig;

        this.errors$ = ValueObservable
            .combineLatestTupleArr(this.inputs.map(input => input.state.errorForForm$))
            .map(filterNull)
        ;

        this.data$ = ValueObservable
            .combineLatestTupleArr(this.inputs.map(input => input.state.value$))
        ;

        this.submitData$ = this.send$
            .withLatestFrom2(
                this.data$,
                this.errors$
            )
            .filter(([click, data, errors]) => errors.length === 0)
            .map(([click, data, errors]) => data);
    }
}
