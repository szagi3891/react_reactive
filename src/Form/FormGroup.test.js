import FormInputState from './FormInputState';
import FormGroupState from './FormGroupState';
import Validators from './Validators';

const getValue = (obs) => {
    let value = null;
    let isSet = false;

    obs.subscribe((newValue) => {
        value = newValue;
        isSet = true;
    });

    if (isSet !== true) {
        throw Error('Nieprawidłowe zachowanie ValueObservable');
    }

    return value;
}

const getStateInput = (inputState) => ({
    errorForInput: getValue(inputState.errorForInput$),
    errorForForm: getValue(inputState.errorForForm$),
    value: getValue(inputState.value$)
});

const getStateForm = (formState) => ({
    inputs: [
        getStateInput(formState.inputs[0].state),
        getStateInput(formState.inputs[1].state),
        getStateInput(formState.inputs[2].state)
    ],
    data: getValue(formState.data$),
    erross: getValue(formState.errors$)
});

const ErrorLable1 = 'Oczekiwano poprawnej daty';
const ErrorLable2 = 'Oczekiwano poprawnego wieku';
const ErrorLable3 = 'Oczekiwano hasła do biosu';

describe('aa', () => {
    
    let formState;

    beforeEach(() => {
        formState = new FormGroupState([{
            key: 'field1',
            label: 'Wprowadź datę bitwy pod Grunwaldem',
            state: new FormInputState(ErrorLable1, Validators.isGrunwald)
        }, {
            key: 'field2',
            label: 'Wprowadź wiek jakiśtam',
            state: new FormInputState(ErrorLable2, Validators.isNumber)
        }, {
            key: 'field3',
            label: 'Wprowadź liczbę szesnastkową',
            state: new FormInputState(ErrorLable3, Validators.isHex)
        }]);
    });

    it('check blur for input', () => {
        const input1$ = formState.inputs[0].state;
        const input2$ = formState.inputs[1].state;
        const input3$ = formState.inputs[2].state;

        expect(getStateForm(formState)).toMatchSnapshot();

        input1$.onBlur();

        expect(getStateForm(formState)).toMatchSnapshot();

        input1$.onChange({
            target: {
                value: '1322'
            }
        });

        expect(getStateForm(formState)).toMatchSnapshot();

        input1$.onChange({
            target: {
                value: '1410'
            }
        });

        expect(getStateForm(formState)).toMatchSnapshot();        
    });
});