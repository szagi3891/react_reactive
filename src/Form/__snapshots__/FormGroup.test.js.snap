// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`aa check blur for input 1`] = `
Object {
  "data": Array [
    "",
    "",
    "",
  ],
  "erross": Array [
    "Oczekiwano poprawnej daty",
    "Oczekiwano poprawnego wieku",
    "Oczekiwano hasła do biosu",
  ],
  "inputs": Array [
    Object {
      "errorForForm": "Oczekiwano poprawnej daty",
      "errorForInput": null,
      "value": "",
    },
    Object {
      "errorForForm": "Oczekiwano poprawnego wieku",
      "errorForInput": null,
      "value": "",
    },
    Object {
      "errorForForm": "Oczekiwano hasła do biosu",
      "errorForInput": null,
      "value": "",
    },
  ],
}
`;

exports[`aa check blur for input 2`] = `
Object {
  "data": Array [
    "",
    "",
    "",
  ],
  "erross": Array [
    "Oczekiwano poprawnej daty",
    "Oczekiwano poprawnego wieku",
    "Oczekiwano hasła do biosu",
  ],
  "inputs": Array [
    Object {
      "errorForForm": "Oczekiwano poprawnej daty",
      "errorForInput": "Oczekiwano poprawnej daty",
      "value": "",
    },
    Object {
      "errorForForm": "Oczekiwano poprawnego wieku",
      "errorForInput": null,
      "value": "",
    },
    Object {
      "errorForForm": "Oczekiwano hasła do biosu",
      "errorForInput": null,
      "value": "",
    },
  ],
}
`;

exports[`aa check blur for input 3`] = `
Object {
  "data": Array [
    "1322",
    "",
    "",
  ],
  "erross": Array [
    "Oczekiwano poprawnej daty",
    "Oczekiwano poprawnego wieku",
    "Oczekiwano hasła do biosu",
  ],
  "inputs": Array [
    Object {
      "errorForForm": "Oczekiwano poprawnej daty",
      "errorForInput": "Oczekiwano poprawnej daty",
      "value": "1322",
    },
    Object {
      "errorForForm": "Oczekiwano poprawnego wieku",
      "errorForInput": null,
      "value": "",
    },
    Object {
      "errorForForm": "Oczekiwano hasła do biosu",
      "errorForInput": null,
      "value": "",
    },
  ],
}
`;

exports[`aa check blur for input 4`] = `
Object {
  "data": Array [
    "1410",
    "",
    "",
  ],
  "erross": Array [
    "Oczekiwano poprawnego wieku",
    "Oczekiwano hasła do biosu",
  ],
  "inputs": Array [
    Object {
      "errorForForm": null,
      "errorForInput": null,
      "value": "1410",
    },
    Object {
      "errorForForm": "Oczekiwano poprawnego wieku",
      "errorForInput": null,
      "value": "",
    },
    Object {
      "errorForForm": "Oczekiwano hasła do biosu",
      "errorForInput": null,
      "value": "",
    },
  ],
}
`;
